XMonad/Layout/Inspect.hs

```diff
@@ -36,23 +38,58 @@ import XMonad
 import qualified XMonad.StackSet as W
 
 -- $usage
+-- This module provides a way for layout authors to make workspace layout data
+-- accessible to end-users.
+--
+-- Best explained by example, suppose you've written a layout modifier:
+--
+-- > newtype Foo a = Foo String deriving (Read, Show)
+-- > instance LayoutModifier Foo a
+-- > foo :: LayoutClass l a => String -> l a -> ModifiedLayout Foo l a
+-- > foo = ModifiedLayout . Foo
+--
+-- To allow users to inspect the string contained in a given workspace's layout,
+-- first import this module:
+--
+-- > import XMonad.Layout.Inspect
+--
+-- Then define instances for 'InspectResult' and 'InspectLayout', and implement
+-- getFoo in terms of 'inspectWorkspace':
+--
+-- > data GetFoo = GetFoo
+-- > type instance InspectResult GetFoo = Alt Maybe String
+-- >
+-- > instance InspectLayout GetFoo Foo a where
+-- >   inspectLayout GetFoo (Foo s) = pure s
+-- >
+-- > getFoo :: (LayoutClass l Window, InspectLayout GetFoo l Window)
+-- >        => l Window -> WindowSpace -> Maybe String
+-- > getFoo l = getAlt . inspectWorkspace l GetFoo
+--
+-- An end-user can then call getFoo by passing it the appropriate layout and a
+-- given workspace:
 --
--- TODO: help for users
+-- > xFoo :: WindowSpace -> X ()
+-- > xFoo wsp = do
+-- >   l <- asks (layoutHook . config)
+-- >   case getFoo l wsp of
+-- >     Nothing -> pure ()
+-- >     Just s -> xmessage s
```

Comment:
@liskin @mislavzanic I encountered some unexpected behavior while testing this. I captured my test setup [here](https://github.com/ivanbrennan/xmonad-testing/commit/9218dee37c0ee35f076de0971bbc9842c6cd51d6), and the most relevant part is [this section](https://github.com/ivanbrennan/xmonad-testing/commit/9218dee37c0ee35f076de0971bbc9842c6cd51d6#diff-60b34e6911d2bed9d2dc5c73ec52011fc2a2baac629ec1fb217a90719a9d7696R32-R51).

In brief, I created a simple layout modifier:
```haskell
newtype Foo a = Foo String deriving (Read, Show)
instance LayoutModifier Foo a

foo :: LayoutClass l a => String -> l a -> ModifiedLayout Foo l a
foo = ModifiedLayout . Foo

data GetFoo = GetFoo
type instance InspectResult GetFoo = Alt Maybe String

instance InspectLayout GetFoo Foo a where
  inspectLayout GetFoo (Foo s) = pure s

getFoo :: (LayoutClass l Window, InspectLayout GetFoo l Window)
       => l Window -> WindowSpace -> Maybe String
getFoo l = getAlt . inspectWorkspace l GetFoo
```
and used it in a `layoutHook`:
```haskell
myLayoutHook = foo "test" (Tall nmaster delta ratio)
  where
     nmaster = 1
     ratio   = 1/2
     delta   = 3/100

main = xmonad $ def { keys = myKeys, layoutHook = myLayoutHook }
```

If I pass `myLayoutHook` as the first argument, `getFoo` returns `Just "test"` as expected. E.g.
```haskell
do
  wsp <- withWindowSet (pure . W.workspace . W.current)
  getFoo myLayoutHook wsp
  -- => Just "test"
```
However, if I retrieve the layout hook by other means, such as `asks`, then `getFoo` returns `Nothing`:
```haskell
do
  wsp <- withWindowSet (pure . W.workspace . W.current)
  l <- asks (layoutHook . config)
  getFoo l wsp
  -- => Nothing
```

Any idea what I'm missing?

Reply:
The `layoutHook` you pull out of the `config` record at runtime is wrapped in the `Layout` existential. However, the actual purpose of the `l a` argument to `inspectWorkspace` and co. is to unwrap that existential by casting its contents to the supplied type, so giving it the wrapped layout doesn't help at all; the cast should fail with a call to `error`.

The `OVERLAPPABLE` instance for `InspectLayout` is then selected, narrowly preventing the `error` from blowing up in your face, relegating it to a `Nothing`.

Reply:
Probably not, as `myLayoutHook` is AFAIK the only place where the actual layout type (not wrapped in an existential) is present

Reply:
It's present in the `XConfig` as well, so we could possibly add something like an `inspectableLayout :: Proxy (l Window) → XConfig l → XConfig l` and then have people do

```haskell
main = xmonad $ inspectableLayout $ \il -> def{ layoutHook = myLayoutHook, … }
```

That might be marginally less error-prone, as some `XConfig` combinators like [`withEasySB`](https://xmonad.github.io/xmonad-docs/xmonad-contrib/XMonad-Hooks-StatusBar.html#v:withEasySB) change the layout, we'd just need to tell people that `inspectableLayout` needs to be the very last. But they'd need to pass the `il` as a parameter every time they need it outside of `main`…

Reply:
Yeah, I really think you'd want to be capturing the layout type immediately before `xmonad` eats it. I suggest something like:

```haskell
-- > main = xmonadIL \i@Inspect{current,tag,workspace} ->
-- >   ...
xmonadIL
  :: (LayoutClass l Window, Read (l Window))
  => (Inspect l -> XConfig l) -> IO ()
xmonadIL = xmonad . inspectableLayout

-- > main = xmonad $ inspectableLayout \i@Inspect{current,tag,workspace} ->
-- >   ...
inspectableLayout
  :: forall l. Typeable l
  => (Inspect l -> XConfig l) -> XConfig l
inspectableLayout k = k (makeInspect (Proxy @l))

data Inspect l = Inspect
  { current   :: forall i. InspectLayout i l Window
              => i                -> X (       InspectResult i )
  , tag       :: forall i. InspectLayout i l Window
              => i -> WorkspaceId -> X (Maybe (InspectResult i))
  , workspace :: forall i. InspectLayout i l Window
              => i -> WindowSpace ->           InspectResult i
  }

makeInspect :: Typeable l => proxy l -> Inspect l
makeInspect p = Inspect
  { current   = inspectCurrentP   p
  , tag       = inspectTagP       p
  , workspace = inspectWorkspaceP p
  }
```

where

```haskell
inspectCurrentP :: (Typeable l, InspectLayout i l Window)
                => proxy l -> i -> X (InspectResult i)
inspectCurrentP p i = gets (inspectWorkspaceP p i . w)
    where w = W.workspace . W.current . windowset

inspectTagP :: (Typeable l, InspectLayout i l Window)
            => proxy l -> i -> WorkspaceId
            -> X (Maybe (InspectResult i))
inspectTagP p i t = gets (fmap (inspectWorkspaceP p i) . mw)
    where mw = find ((t==) . W.tag) . W.workspaces . windowset

inspectWorkspaceP :: (Typeable l, InspectLayout i l Window)
                  => proxy l -> i -> WindowSpace -> InspectResult i
inspectWorkspaceP p i = inspectLayout i . asLayout p . W.layout

asLayout :: (Typeable l, Typeable a) => proxy l -> Layout a -> l a
asLayout _ (Layout l) = cast' l
```

and

```haskell
inspectCurrent :: forall l i. (Typeable l, InspectLayout i l Window)
               => l Window -> i -> X (InspectResult i)
inspectCurrent _ = inspectCurrentP (Proxy @l)

inspectTag :: forall l i. (Typeable l, InspectLayout i l Window)
           => l Window -> i -> WorkspaceId
           -> X (Maybe (InspectResult i))
inspectTag _ = inspectTagP (Proxy @l)

inspectWorkspace :: forall l i. (Typeable l, InspectLayout i l Window)
                 => l Window -> i -> WindowSpace -> InspectResult i
inspectWorkspace _ = inspectWorkspaceP (Proxy @l)
```

It can't really be helped that `i` needs to be passed around to be used outside of `main`; this is the only place we know the actual layout type. The alternative is to tell users to write their configs like:

```haskell
main = xmonad myConfig

myConfig = ewmh . ... $ def
  { ...
  }
```

and use e.g. `inspectCurrentP myConfig`. Of if IO is necessary:

```haskell
main = makeMyConfig >>= xmonad

makeMyConfig :: IO (XConfig L)
makeMyConfig = do
  ...
```

using `inspectCurrentP (Compose makeMyConfig)`.